src/compiler/core.ts

```diff
@@ -661,6 +669,8 @@ namespace ts {
     /**
      * Compacts an array, removing any falsey elements.
      */
+    export function compact<T>(array: T[]): T[];
+    export function compact<T>(array: ReadonlyArray<T>): ReadonlyArray<T>;

```

Comment:
The more-specific overload should go first -- `T[]` is more specific than `ReadonlyArray` since it contains everything `ReadonlyArray` has, plus mutators.